libraries/ms-common/src/can_pack_impl.c

```diff
@@ -0,0 +1,53 @@
+#include "can_pack_impl.h"
+
+#include <limits.h>
+#include <stdint.h>
+
+#include "can_msg.h"
+
+// Calculate the number of used bytes if the trailing bytes are 0 or unused then there is no point
+// wasting bus time transmitting them.
+#define CAN_PACK_CALC_DLC(data) CHAR_BIT - (size_t)__builtin_ffsll((int64_t)(data)) / CHAR_BIT
```

Comment:
This seems a little odd. Shouldn't it be the other way around? If we're using bytes 0, 1, 2 they'll be at the leftmost side. You should be counting leading 0s afaik? I think it'd be nicer to explicitly set it using a length field.

Reply:
? Sorry not following... Let me explain then maybe you can point out the error in my logic:

So CHAR_BIT = 8 so we have 8 - (empty bits / 8) and we use bytes 0, 1, 2 will give 5 since counting trailing 0s and dividing by 8 will give the empty bytes on the right. Then we remove that from 8 to get the number of bytes used? 

As for whether this is should be a field I'm inclined to disagree? If I'm writing the application layer I don't want to worry about what size of message I want to send. This computes it efficiently and IMO can be more efficient since trailing 0 bytes can just be elided automatically without having to think about it. Although perhaps I'm missing something important that can be done by setting it explicitly?

Reply:
```
  union {
    uint64_t raw;
    uint8_t u8[8];
  } data;

  data.u8[0] = 0xAA;
  printf("0x%lx - ffsll: %d (%lu)\n", data.raw, __builtin_ffsll((int64_t)data.raw),
         CHAR_BIT - (size_t)__builtin_ffsll((int64_t)data.raw) / CHAR_BIT);
```
Prints `0xaa - ffsll: 2 (8)`

Reply:
re: DLC: If you have leading zeros as a parameter (or trailing zeros, depending on what you're cutting out), your TX and RX DLCs will be different.

example:
```
TX of f1 = 0x00
RX of DLC = 0
```

Technically, this will still work with your pack/unpack functions since I do memset everything to zero, but I think doing a basic length check is useful. It'll make sure we don't somehow mess something up with malformed packets. For example, connecting an old board with different CAN IDs to the network.

Reply:
I think I'll just add a `num_bytes` field and then add the DLC population in codegen since it already knows how many fields to send. It makes life easier.

Reply:
Done.

Reply:
re: DLC, that's what I meant. Sorry if you misunderstood me.